File: src/main/java/no/nav/bidrag/reisekostnad/tjeneste/Arkiveringstjeneste.java
package no.nav.bidrag.reisekostnad.tjeneste;

import java.time.LocalDateTime;
import javax.transaction.Transactional;
import no.nav.bidrag.reisekostnad.database.datamodell.Forespørsel;
import no.nav.bidrag.reisekostnad.integrasjon.bidrag.doument.BidragDokumentkonsument;
import no.nav.bidrag.reisekostnad.integrasjon.bidrag.doument.pdf.PdfGenerator;
import no.nav.bidrag.reisekostnad.tjeneste.støtte.Mapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class Arkiveringstjeneste {

  private final static String REISEKOSTNAD_REFERANSEIDPREFIKS = "REISEKOSTNAD_";
  private final BidragDokumentkonsument bidragDokumentkonsument;
  private final Mapper mapper;
  private final Databasetjeneste databasetjeneste;
  private final PdfGenerator pdfGenerator;

  @Autowired
  public Arkiveringstjeneste(
      BidragDokumentkonsument bidragDokumentkonsument,
      Mapper mapper,
      Databasetjeneste databasetjeneste,
      PdfGenerator pdfGenerator) {
    this.bidragDokumentkonsument = bidragDokumentkonsument;
    this.mapper = mapper;
    this.databasetjeneste = databasetjeneste;
    this.pdfGenerator = pdfGenerator;
  }

  @Transactional
  public String arkivereForespørsel(int idForespørsel) {
    var forespørsel = databasetjeneste.henteAktivForespørsel(idForespørsel);
    var respons = bidragDokumentkonsument.oppretteJournalpost(forespørsel.getHovedpart().getPersonident(),
        REISEKOSTNAD_REFERANSEIDPREFIKS + idForespørsel);

    forespørsel.setJournalført(LocalDateTime.now());
    forespørsel.setIdJournalpost(respons.getJournalpostId());

    return respons.getJournalpostId();
  }

  private byte[] opprettePdf(Forespørsel forespørsel) {

    var barn = mapper.tilPersonDto(forespørsel.getBarn());
    var hovedpart = mapper.tilPersonDto(forespørsel.getHovedpart().getPersonident());
    var motpart = mapper.tilPersonDto(forespørsel.getMotpart().getPersonident());

    return pdfGenerator.genererePdf(barn, hovedpart, motpart);
  }
}

File: src/main/java/no/nav/bidrag/reisekostnad/tjeneste/Arkiveringstjeneste.kt
package no.nav.bidrag.reisekostnad.tjeneste

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.transaction.Transactional
import no.nav.bidrag.reisekostnad.database.datamodell.Forespørsel
import no.nav.bidrag.reisekostnad.feilhåndtering.Feilkode
import no.nav.bidrag.reisekostnad.feilhåndtering.Valideringsfeil
import no.nav.bidrag.reisekostnad.integrasjon.bidrag.doument.BidragDokumentkonsument
import no.nav.bidrag.reisekostnad.integrasjon.bidrag.doument.pdf.PdfGenerator
import no.nav.bidrag.reisekostnad.model.erArkivert
import no.nav.bidrag.reisekostnad.model.kanArkiveres
import no.nav.bidrag.reisekostnad.tjeneste.støtte.Mapper
import org.springframework.stereotype.Service
import java.time.LocalDateTime


private val log = KotlinLogging.logger {}

@Service
class Arkiveringstjeneste(
    val bidragDokumentkonsument: BidragDokumentkonsument,
    val mapper: Mapper,
    val databasetjeneste: Databasetjeneste
) {
    @Transactional
    fun arkivereForespørsel(idForespørsel: Int) {
        try {
            val forespørsel = databasetjeneste.henteAktivForespørsel(idForespørsel)

            if (forespørsel.erArkivert) return
            if (!forespørsel.kanArkiveres) throw Valideringsfeil(Feilkode.KAN_IKKE_ARKIVERE_FORESPØRSEL)

            val pdfDokument = opprettPdf(forespørsel)
            val referanseId = "$REISEKOSTNAD_REFERANSEIDPREFIKS$idForespørsel"

            val respons = bidragDokumentkonsument.opprettJournalpost(forespørsel.hovedpart.personident, referanseId, pdfDokument)
            forespørsel.journalført = LocalDateTime.now()
            forespørsel.idJournalpost = respons.journalpostId
            log.info { "Arkivert dokument for forespørsel $idForespørsel med journalpostid ${respons.journalpostId}" }
        } catch (e: Exception){
            log.error("Det skjedde en feil ved arkivering av dokument for forespørsel $idForespørsel", e)
        }
    }

    private fun opprettPdf(forespørsel: Forespørsel): ByteArray {
        val barn = forespørsel.barn.map { mapper.tilPersonDto(it?.personident) }.toSet()
        val hovedpart = mapper.tilPersonDto(forespørsel.hovedpart.personident)
        val motpart = mapper.tilPersonDto(forespørsel.motpart.personident)
        return PdfGenerator.genererePdf(barn, hovedpart, motpart, forespørsel.samtykket)
    }

    companion object {
        private const val REISEKOSTNAD_REFERANSEIDPREFIKS = "REISEKOSTNAD_"
    }
}